File: frontend/mcm-app/src/bff-server/audit-sink.ts
/**
 * BFF audit sink (T076c) — config-gated, append-only OpenSearch sink.
 *
 * ALWAYS emits `logger.audit(...)` (today's behaviour, unchanged).
 * ALSO, when OPENSEARCH_URL is set, fire-and-forget appends a redacted doc to OpenSearch.
 *
 * Hard constraints:
 *   - Fire-and-forget: the OpenSearch POST is NEVER awaited on the response path.
 *   - Never throws: all errors are swallowed with `.catch` so the route is unaffected.
 *   - Never sends PII or credentials: `buildAuditDoc` strips the same sensitive keys the
 *     BFF logger redacts before the doc is POSTed.
 *   - Additive only: OPENSEARCH_URL unset ⇒ behaviour is exactly today's (log only).
 */

import { logger } from '@/bff-server/logger';

/**
 * Keys that must never appear in the OpenSearch document.
 * Mirrors the BFF logger's SENSITIVE_KEYS set, plus a catch-all for any key
 * whose name contains "token" (case-insensitive).
 */
const REDACT = new Set([
  'token',
  'sessionId', 'session_id',
  'password',
  'secret', 'clientSecret', 'client_secret',
  'cookie',
  'authorization',
  'code', 'codeVerifier', 'code_verifier',
  'email',
  'username',
  // Explicit token variants (also caught by the .includes('token') check below).
  'accessToken', 'refreshToken', 'idToken',
  'id_token', 'access_token', 'refresh_token',
]);

/**
 * Build a redacted audit document safe for external storage.
 * Strips every key in REDACT and every key whose lowercased name contains "token".
 * The `action` field is always included.
 */
export function buildAuditDoc(
  action: string,
  fields: Record<string, unknown>,
): Record<string, unknown> {
  const doc: Record<string, unknown> = { action };
  for (const [k, v] of Object.entries(fields)) {
    if (REDACT.has(k) || k.toLowerCase().includes('token')) continue;
    doc[k] = v;
  }
  return doc;
}

/**
 * Always emits the structured audit log via `logger.audit` (synchronous, original fields
 * passed verbatim — the logger applies its own redaction).
 *
 * When OPENSEARCH_URL is set, ALSO fire-and-forget appends a redacted doc to OpenSearch.
 * The POST is never awaited, never throws, and never delays the route response.
 */
export function audit(action: string, fields: Record<string, unknown>): void {
  // Always log — original fields, logger does its own redaction.
  logger.audit(action, fields);

  const base = (process.env['OPENSEARCH_URL'] ?? '').trim();
  if (!base) return;

  const username = process.env['OPENSEARCH_USERNAME'] ?? '';
  const password = process.env['OPENSEARCH_PASSWORD'] ?? '';
  const auth = 'Basic ' + Buffer.from(`${username}:${password}`).toString('base64');
  const url = `${base.replace(/\/$/, '')}/mcm-agent-audit/_doc`;

  // Fire-and-forget — void the promise so it is never awaited on the response path.
  // The .catch ensures this can never throw or cause an unhandled rejection.
  void fetch(url, {
    method: 'POST',
    headers: {
      'content-type': 'application/json',
      authorization: auth,
    },
    body: JSON.stringify(buildAuditDoc(action, fields)),
  }).catch((err: unknown) =>
    logger.error('audit append failed', { action: 'audit_append', error: err }),
  );
}

File: agents/movie-assistant/src/audit_sink.py
"""Append-only audit sink for the agent gateway (T076b).

Two behaviours, always additive (SC-005):
- **Always**: structured log via the module logger (today's behaviour, no change).
- **When OPENSEARCH_URL is set**: also best-effort POST an audit document to the
  `mcm-agent-audit` index. The POST is *awaited* but bounded to a 3-second timeout so the
  tool path never waits longer than that, and any exception is swallowed — audit is
  never-blocking, never-raising from the caller's perspective.

Non-blocking approach chosen: **fire-and-forget via `asyncio.ensure_future`**.
Call sites in `src/tools/mcp_tools.py` schedule `emit_audit` as a background task using
`asyncio.ensure_future` — the coroutine is never awaited by the caller. Internally,
`emit_audit` itself awaits the HTTP POST but bounds it to a 3-second timeout so the
worst-case stall is predictable and far shorter than any user-visible timeout. The
`except Exception` block ensures even a timeout or connection error does not propagate.

SC-004 / token-leak compliance: `build_audit_doc` strips every key whose lower-cased name
contains any of the TOKEN_MARKERS before the log line and before the POST. No token-named
variable is ever passed to a logger call in this module.
"""

from __future__ import annotations

import logging
import os
from collections.abc import Mapping
from typing import Any

import httpx

logger = logging.getLogger(__name__)

# Mirror the markers from src/eval/token_leak_scan.TOKEN_MARKERS so the same identifiers
# are treated as sensitive here and in the static scan.
_REDACT_MARKERS: tuple[str, ...] = (
    "token",
    "jwt",
    "authorization",
    "bearer",
    "secret",
    "password",
    "credential",
    "email",
    "username",
    "cookie",
    "code",
)


def build_audit_doc(action: str, fields: Mapping[str, Any]) -> dict[str, Any]:
    """Return ``{**redacted_fields, "action": action}`` — pure, no side effects.

    Any key whose lower-cased name contains a marker from ``_REDACT_MARKERS`` is dropped.
    The ``action`` key is always present and is set to the supplied value.
    """
    doc: dict[str, Any] = {
        k: v
        for k, v in fields.items()
        if not any(marker in k.lower() for marker in _REDACT_MARKERS)
    }
    doc["action"] = action
    return doc


async def emit_audit(
    action: str,
    fields: Mapping[str, Any],
    *,
    env: Mapping[str, str] | None = None,
    client: httpx.AsyncClient | None = None,
) -> None:
    """Emit a structured audit log line and, when ``OPENSEARCH_URL`` is set, POST to OpenSearch.

    Always logs via the module logger (no sensitive keys — ``build_audit_doc`` redacts first).
    When ``OPENSEARCH_URL`` is set, also POSTs the redacted document to the
    ``mcm-agent-audit/_doc`` endpoint with a 3-second timeout.  Any exception from the POST
    is caught and logged at WARNING level — the audit must never raise or stall the caller.

    Parameters
    ----------
    action:
        Audit action label (e.g. ``"agent_tool_call"``).
    fields:
        Arbitrary key→value pairs to include.  Sensitive keys are stripped by
        ``build_audit_doc`` before logging or sending.
    env:
        Environment mapping (defaults to ``os.environ``).  Injected in tests.
    client:
        An ``httpx.AsyncClient`` to use for the POST.  When *None* a temporary client is
        created and closed in this call.  Injected in tests to avoid real network calls.
    """
    e: Mapping[str, str] = os.environ if env is None else env
    doc = build_audit_doc(action, fields)

    # Structured log — only the redacted doc; no raw fields, no token-named variables.
    logger.info(
        "audit %s agent=%s tool=%s status=%s",
        action,
        doc.get("agent"),
        doc.get("tool"),
        doc.get("status"),
    )

    base = (e.get("OPENSEARCH_URL") or "").strip()
    if not base:
        return

    url = f"{base.rstrip('/')}/mcm-agent-audit/_doc"
    auth = (e.get("OPENSEARCH_USERNAME", ""), e.get("OPENSEARCH_PASSWORD", ""))
    try:
        if client is not None:
            await client.post(url, json=doc, auth=auth)
        else:
            http = httpx.AsyncClient(verify=False, timeout=3.0)  # noqa: S501
            try:
                await http.post(url, json=doc, auth=auth)
            finally:
                await http.aclose()
    except Exception:  # best-effort — audit must never break the call  # noqa: BLE001
        logger.warning("audit append to OpenSearch failed", exc_info=False)
